### PR TITLE
[AMDGPU] Let LowerModuleLDS run twice on the same module

### PR DESCRIPTION
**Pull request by Pierre-vh**
If all variables in the module are absolute, this means we're running the pass again on an already lowered module, and that works.
If none of them are absolute, lowering can proceed as usual.
Only diagnose cases where we have a mix of absolute/non-absolute GVs, which means we added LDS GVs after lowering, which is broken.

See #81491
Split from #75333

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp [hunk 1]
```diff
@@ -327,16 +327,18 @@ class AMDGPULowerModuleLDS {
     return convertUsersOfConstantsToInstructions(LDSGlobals);
   }
 
+  std::optional<bool> HasAbsoluteGVs;
```
Comment (JonChesterfield) [inline llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp:330]: This is only used by one function, make it a local variable and reinstate the static annotations. Minimising scope of mutable state is important.

File: llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp [hunk 2]
```diff
@@ -345,10 +347,18 @@ class AMDGPULowerModuleLDS {
         continue;
       }
 
-      if (GV.isAbsoluteSymbolRef()) {
-        report_fatal_error(
-            "LDS variables with absolute addresses are unimplemented.");
-      }
+      // Check if the module is consistent: either all GVs are absolute (happens
+      // when we run the pass more than once), or none are.
+      if (HasAbsoluteGVs.has_value()) {
```
Comment (JonChesterfield) [inline llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp:352]: Lets call GV.isAbsoluteSymbolRef once instead of three times

**Review by JonChesterfield — CHANGES_REQUESTED** [review]

**Comment by Pierre-vh** [comment]
ping

**Review by JonChesterfield — APPROVED** [review]
Thanks for the cleanup. Latest commit looks good to me and risk profile on it is low, let's ship it. Thanks!

**Pull request merged**